In [1]:
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import pandas as pd
import statsmodels.api as sm
import statsmodels.tsa.api as smt
import statsmodels.formula.api as smf
import pylab as py 

from sklearn import linear_model
from sqlalchemy import create_engine
from statsmodels.stats.outliers_influence import variance_inflation_factor

In [2]:
# Display preferences.
%matplotlib inline
pd.options.display.float_format = '{:.3f}'.format

import warnings
warnings.filterwarnings(action="ignore")

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'weatherinszeged'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

weather = pd.read_sql_query('select * from weatherinszeged',con=engine)

# no need for an open connection, as we're only doing a single query
engine.dispose()

weather.head(10)

date        summary preciptype  temperature  \
0 2006-03-31 22:00:00+00:00  Partly Cloudy       rain        9.472   
1 2006-03-31 23:00:00+00:00  Partly Cloudy       rain        9.356   
2 2006-04-01 00:00:00+00:00  Mostly Cloudy       rain        9.378   
3 2006-04-01 01:00:00+00:00  Partly Cloudy       rain        8.289   
4 2006-04-01 02:00:00+00:00  Mostly Cloudy       rain        8.756   
5 2006-04-01 03:00:00+00:00  Partly Cloudy       rain        9.222   
6 2006-04-01 04:00:00+00:00  Partly Cloudy       rain        7.733   
7 2006-04-01 05:00:00+00:00  Partly Cloudy       rain        8.772   
8 2006-04-01 06:00:00+00:00  Partly Cloudy       rain       10.822   
9 2006-04-01 07:00:00+00:00  Partly Cloudy       rain       13.772   

   apparenttemperature  humidity  windspeed  windbearing  visibility  \
0                7.389     0.890     14.120      251.000      15.826   
1                7.228     0.860     14.265      259.000      15.826   
2                9.378     0.890      3.928      204.000      14.957   
3                5.944     0.830     14.104      269.000      15.826   
4                6.978     0.830     11.045      259.000      15.826   
5                7.111     0.850     13.959      258.000      14.957   
6                5.522     0.950     12.365      259.000       9.982   
7                6.528     0.890     14.152      260.000       9.982   
8               10.822     0.820     11.318      259.000       9.982   
9               13.772     0.720     12.526      279.000       9.982   

   loudcover  pressure                       dailysummary  
0      0.000  1015.130  Partly cloudy throughout the day.  
1      0.000  1015.630  Partly cloudy throughout the day.  
2      0.000  1015.940  Partly cloudy throughout the day.  
3      0.000  1016.410  Partly cloudy throughout the day.  
4      0.000  1016.510  Partly cloudy throughout the day.  
5      0.000  1016.660  Partly cloudy throughout the day.  
6      0.000  1016.720  Partly cloudy throughout the day.  
7      0.000  1016.840  Partly cloudy throughout the day.  
8      0.000  1017.370  Partly cloudy throughout the day.  
9      0.000  1017.220  Partly cloudy throughout the day.

In [3]:
Y = weather['apparenttemperature'] - weather['temperature']


# target variable apprentteperature - teperature
#Y = weather - [["apparenttemperature", "temperature"]]

# Feature variables
X = weather[["humidity", "windspeed"]]

#X_const = weather [["humidity", "windspeed"]]

# We create a LinearRegression model object
# from scikit-learn's linear_model module.
lrm = linear_model.LinearRegression()

# fit method estimates the coefficients using OLS
lrm.fit(X, Y)

# Inspect the results.
print('\nCoefficients: \n', lrm.coef_)
print('\nIntercept: \n', lrm.intercept_)


Coefficients: 
 [-3.02918594 -0.11929075]

Intercept: 
 2.4381054151877004


In [5]:
X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Thu, 06 Feb 2020   Prob (F-statistic):               0.00
Time:                        17:11:08   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.4381      0.021    115.948      0.000       2.397       2.479
humidity      -3.0292      0.024   -126.479      0.000      -3.076      -2.982
windspeed     -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.267
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

# What are the R-squared and adjusted R-squared values? Do you think they are satisfactory? Why?

The R-squared and adjusted R-squared values are the same they both equal 0.288 in this case the R-squared and adjusted R-squared only explains 28% of the target variable. The values are not satisfactory.

# Next, include the interaction of humidity and windspeed to the model above and estimate the model using OLS. Now, what is the R-squared of this model? Does this model improve upon the previous one?

In [6]:
weather["humidity_windspeed_interaction"] = weather.humidity * weather.windspeed

# Y is the target variable
Y = weather["apparenttemperature"] - weather["temperature"]
# X is the feature set
X = weather[["humidity", "windspeed", "humidity_windspeed_interaction"]]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Thu, 06 Feb 2020   Prob (F-statistic):               0.00
Time:                        17:20:05   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==================================================================================================
                                     coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
const                              0.0839      0.033      2.511      0.012       0.018       0.149
humidity                           0.1775      0.043      4.133      0.000       0.093       0.262
windspeed                          0.0905      0.002     36.797      0.000       0.086       0.095
humidity_windspeed_interaction    -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.265
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The R-squared and adjusted R-squared values are the same they both equal 0.341 in this case the R-squared and adjusted R-squared only explains 34% of the target variable. The interaction of "humidity_windspeed_interaction" explains some previously unexplained variance in the temperature. This model is an improve of the previous one becaue the value increased by 0.053 points.

# Add visibility as an additional explanatory variable to the first model and estimate it. Did R-squared increase? What about adjusted R-squared? Compare the differences put on the table by the interaction term and the visibility in terms of the improvement in the adjusted R-squared. Which one is more useful?

In [9]:
# Y is the target variable
Y = weather["apparenttemperature"] - weather["temperature"]
# X is the feature set with added visibility
X = weather[["humidity","windspeed", "visibility"]]

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Thu, 06 Feb 2020   Prob (F-statistic):               0.00
Time:                        17:53:09   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.5756      0.028     56.605      0.000       1.521       1.630
humidity      -2.6066      0.025   -102.784      0.000      -2.656      -2.557
windspeed     -0.1199      0.001   -179.014      0.000      -0.121      -0.119
visibility     0.0540      0.001     46.614      0.000       0.052       0.056
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.282
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                          -0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

From this summary the R-squared is 0,304 and adjusted R- squared is 0.303. R-squared and adjusted R-squared increased compared to the very first model. visibility helped improved the adjusted R-squared by 0.015 point. This improvement is lower than the intersetion model. The intersection model seems to be critical to and better than the other two models.

# Choose the best one from the three models above with respect to their AIC and BIC scores. Validate your choice by discussing your justification with your mentor.

 The first model's AIC and BIC values are 340900. The second model's are 333400 and the third model's are 338800. Among these, the model with the lowest AIC and BIC scores is the second model. So, the best of these three models is the second one.